File: triad-sample/app/src/test/java/com/nhaarman/triad/sample/editnote/EditNewNotePresenterTest.java
/*
 * Copyright 2016 Niek Haarman
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nhaarman.triad.sample.editnote;

import android.content.Context;
import android.content.res.Resources;
import com.nhaarman.triad.Optional;
import com.nhaarman.triad.Screen;
import com.nhaarman.triad.Triad;
import com.nhaarman.triad.sample.ActivityComponent;
import com.nhaarman.triad.sample.MemoryNoteRepository;
import com.nhaarman.triad.sample.Note;
import com.nhaarman.triad.sample.NoteCreator;
import com.nhaarman.triad.sample.NoteValidator;
import org.junit.Before;
import org.junit.Test;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class EditNewNotePresenterTest {

    private static final String TITLE = "title";

    private static final String CONTENTS = "contents";

    private EditNotePresenter mEditNotePresenter;

    private EditNoteContainer mEditNoteContainerMock;

    private NoteValidator mNoteValidator;

    private NoteCreator mNoteCreatorMock;

    private MemoryNoteRepository mNoteRepository;

    private Triad mTriad;

    @Before
    public void setUp() {
        mNoteRepository = spy(new MemoryNoteRepository());
        mNoteValidator = spy(new NoteValidator());
        mNoteCreatorMock = spy(new NoteCreator(mNoteRepository));

        mTriad = mock(Triad.class);

        mEditNotePresenter = spy(new EditNotePresenter(null, mNoteValidator, mNoteCreatorMock, mNoteRepository, mTriad));
        doReturn(Optional.of(mock(Resources.class))).when(mEditNotePresenter).resources();

        mEditNoteContainerMock = mock(EditNoteContainer.class);
        when(mEditNoteContainerMock.context()).thenReturn(mock(Context.class));
    }

    @Test
    public void onControlGained_setsEmptyTitleToEditNoteContainer() {
    /* Given */
        EditNoteContainer containerMock = mock(EditNoteContainer.class);
        ActivityComponent activityComponentMock = mock(ActivityComponent.class);

    /* When */
        mEditNotePresenter.onControlGained(containerMock, activityComponentMock);

    /* Then */
        verify(containerMock).setTitle("");
    }

    @Test
    public void onControlGained_setsEmptyContentsToEditNoteContainer() {
    /* Given */
        EditNoteContainer containerMock = mock(EditNoteContainer.class);
        ActivityComponent activityComponentMock = mock(ActivityComponent.class);

    /* When */
        mEditNotePresenter.onControlGained(containerMock, activityComponentMock);

    /* Then */
        verify(containerMock).setContents("");
    }

    @Test
    public void onSaveNoteClicked_validatesTitle() {
    /* Given */
        mEditNotePresenter.setContainer(mEditNoteContainerMock);

        when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
        when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
        mEditNotePresenter.onSaveNoteClicked();

    /* Then */
        verify(mNoteValidator).validateTitle(TITLE);
    }

    @Test
    public void onSaveNoteClicked_validatesContents() {
    /* Given */
        mEditNotePresenter.setContainer(mEditNoteContainerMock);

        when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
        when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
        mEditNotePresenter.onSaveNoteClicked();

    /* Then */
        verify(mNoteValidator).validateContents(CONTENTS);
    }

    @Test
    public void onSaveNoteClickedWithEmptyTitle_showsErrorMessage() {
    /* Given */
        mEditNotePresenter.setContainer(mEditNoteContainerMock);

        when(mEditNoteContainerMock.getTitle()).thenReturn("");
        when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
        mEditNotePresenter.onSaveNoteClicked();

    /* Then */
        verify(mEditNoteContainerMock).setTitleError(any(String.class));
    }

    @Test
    public void onSaveNoteClickedWithEmptyContents_showsErrorMessage() {
    /* Given */
        mEditNotePresenter.setContainer(mEditNoteContainerMock);

        when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
        when(mEditNoteContainerMock.getContents()).thenReturn("");

    /* When */
        mEditNotePresenter.onSaveNoteClicked();

    /* Then */
        verify(mEditNoteContainerMock).setContentsError(any(String.class));
    }

    @Test
    public void onSaveNoteClickedWithInvalidProperties_staysOnEditNoteScreen() {
    /* Given */
        mEditNotePresenter.setContainer(mEditNoteContainerMock);

        when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
        when(mEditNoteContainerMock.getContents()).thenReturn("");

    /* When */
        mEditNotePresenter.onSaveNoteClicked();

    /* Then */
        verify(mTriad, never()).goTo(any(Screen.class));
    }

    @Test
    public void onSaveNoteClicked_createsANoteWithTheProperTitle() {
    /* Given */
        mEditNotePresenter.setContainer(mEditNoteContainerMock);

        when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
        when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
        mEditNotePresenter.onSaveNoteClicked();

    /* Then */
        verify(mNoteCreatorMock).createNote(eq(TITLE), any(String.class));
    }

    @Test
    public void onSaveNoteClicked_createsANoteWithTheProperContents() {
    /* Given */
        mEditNotePresenter.setContainer(mEditNoteContainerMock);

        when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
        when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
        mEditNotePresenter.onSaveNoteClicked();

    /* Then */
        verify(mNoteCreatorMock).createNote(any(String.class), eq(CONTENTS));
    }

    @Test
    public void onSaveNoteClicked_persistsTheCreatedNote() {
    /* Given */
        mEditNotePresenter.setContainer(mEditNoteContainerMock);

        when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
        when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
        mEditNotePresenter.onSaveNoteClicked();

    /* Then */
        verify(mNoteRepository).create(any(Note.class));
    }

    @Test
    public void onSaveNoteClicked_navigatesBack() {
    /* Given */
        mEditNotePresenter.setContainer(mEditNoteContainerMock);

        when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
        when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
        mEditNotePresenter.onSaveNoteClicked();

    /* Then */
        verify(mTriad).goBack();
    }
}

File: triad-sample/app-kotlin/src/test/kotlin/com/nhaarman/triad/sample/editnote/EditNewNotePresenterTest.kt
/*
 * Copyright 2016 Niek Haarman
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nhaarman.triad.sample.editnote

import com.nhaarman.mockito.*
import com.nhaarman.triad.Triad
import com.nhaarman.triad.sample.ActivityComponent
import com.nhaarman.triad.sample.MemoryNoteRepository
import com.nhaarman.triad.sample.NoteCreator
import com.nhaarman.triad.sample.NoteValidator
import org.junit.Before
import org.junit.Test


class EditNewNotePresenterTest {

    private lateinit var mEditNotePresenter: EditNotePresenter

    private lateinit var mEditNoteContainerMock: EditNoteContainer

    private lateinit var mNoteValidator: NoteValidator

    private lateinit var mNoteCreatorMock: NoteCreator

    private lateinit var mNoteRepository: MemoryNoteRepository

    private lateinit var mTriad: Triad

    @Before
    fun setUp() {
        mNoteRepository = spy(MemoryNoteRepository())
        mNoteValidator = spy(NoteValidator())
        mNoteCreatorMock = spy(NoteCreator(mNoteRepository))

        mTriad = mock()

        mEditNotePresenter = EditNotePresenter(null, mNoteValidator, mNoteCreatorMock, mNoteRepository, mTriad)

        mEditNoteContainerMock = mock()
        whenever(mEditNoteContainerMock.context()).thenReturn(mock())
    }

    @Test
    fun onControlGained_setsEmptyTitleToEditNoteContainer() {
        /* Given */
        val containerMock = mock<EditNoteContainer>()
        val activityComponentMock = mock<ActivityComponent>()

        /* When */
        mEditNotePresenter.onControlGained(containerMock, activityComponentMock)

        /* Then */
        verify(containerMock).title = ""
    }

    @Test
    fun onControlGained_setsEmptyContentsToEditNoteContainer() {
        /* Given */
        val containerMock = mock<EditNoteContainer>()
        val activityComponentMock = mock<ActivityComponent>()

        /* When */
        mEditNotePresenter.onControlGained(containerMock, activityComponentMock)

        /* Then */
        verify(containerMock).contents = ""
    }

    @Test
    fun onSaveNoteClicked_validatesTitle() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<NoteValidator>(mNoteValidator).validateTitle(TITLE)
    }

    @Test
    fun onSaveNoteClicked_validatesContents() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<NoteValidator>(mNoteValidator).validateContents(CONTENTS)
    }

    @Test
    fun onSaveNoteClickedWithEmptyTitle_showsErrorMessage() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn("")
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<EditNoteContainer>(mEditNoteContainerMock).setTitleError(any())
    }

    @Test
    fun onSaveNoteClickedWithEmptyContents_showsErrorMessage() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn("")

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<EditNoteContainer>(mEditNoteContainerMock).setContentsError(any())
    }

    @Test
    fun onSaveNoteClickedWithInvalidProperties_staysOnEditNoteScreen() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn("")

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<Triad>(mTriad, never()).goTo(any())
    }

    @Test
    fun onSaveNoteClicked_createsANoteWithTheProperTitle() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<NoteCreator>(mNoteCreatorMock).createNote(eq(TITLE), any())
    }

    @Test
    fun onSaveNoteClicked_createsANoteWithTheProperContents() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<NoteCreator>(mNoteCreatorMock).createNote(any(), eq(CONTENTS))
    }

    @Test
    fun onSaveNoteClicked_persistsTheCreatedNote() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<MemoryNoteRepository>(mNoteRepository).create(any())
    }

    @Test
    fun onSaveNoteClicked_navigatesBack() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<Triad>(mTriad).goBack()
    }

    companion object {

        private val TITLE = "title"

        private val CONTENTS = "contents"
    }
}